src/calendar.cpp

```diff
@@ -556,11 +556,24 @@ time_duration rng( time_duration lo, time_duration hi )
 
 season_type season_of_year( const time_point &p )
 {
-    if( calendar::eternal_season() ) {
-        // If we use calendar::start to determine the initial season, and the user shortens the season length
-        // mid-game, the result could be the wrong season!
-        return calendar::initial_season;
+    static time_point prev_turn = calendar::before_time_starts;
+    static season_type prev_season = SPRING;
+    static bool is_eternal = false;
+    
+    if( p != prev_turn ) {
+        prev_turn = p;
+        if( ( is_eternal = calendar::eternal_season() ) ) {
```

Comment:
You can pull this check out and do it first, since you're doing this is both the `if` and the `else`

Reply:
This is assignment as well as flow control. Takes care of setting is_eternal, since the option may have changed since last turn, then decides whether to return the initial season, or proceed to check the current season based on turn.